providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+#include <openssl/err.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <internal/worker.h>
+#include <internal/numbers.h>
+#include <prov/implementations.h>
+#include <prov/provider_ctx.h>
+#include <prov/providercommonerr.h>
+#include <prov/blake2.h>
+
+#ifndef OPENSSL_NO_ARGON2
+
+# define ARGON2_MIN_LANES UINT32_C(1)
+# define ARGON2_MAX_LANES UINT32_C(0xFFFFFF)
+# define ARGON2_MIN_THREADS UINT32_C(1)
+# define ARGON2_MAX_THREADS UINT32_C(0xFFFFFF)
+# define ARGON2_SYNC_POINTS UINT32_C(4)
+# define ARGON2_MIN_OUTLEN UINT32_C(4)
+# define ARGON2_MAX_OUTLEN UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_MEMORY (2 * ARGON2_SYNC_POINTS)
+# define ARGON2_MIN(a, b) ((a) < (b) ? (a) : (b))
+# define ARGON2_MAX_MEMORY_BITS                                                 \
+    ARGON2_MIN(UINT32_C(32), (sizeof(void *) * CHAR_BIT - 10 - 1))
+# define ARGON2_MAX_MEMORY                                                      \
+    ARGON2_MIN(UINT32_C(0xFFFFFFFF), UINT64_C(1) << ARGON2_MAX_MEMORY_BITS)
+# define ARGON2_MIN_TIME UINT32_C(1)
+# define ARGON2_MAX_TIME UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_PWD_LENGTH UINT32_C(0)
+# define ARGON2_MAX_PWD_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_AD_LENGTH UINT32_C(0)
+# define ARGON2_MAX_AD_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_SALT_LENGTH UINT32_C(8)
+# define ARGON2_MAX_SALT_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_SECRET UINT32_C(0)
+# define ARGON2_MAX_SECRET UINT32_C(0xFFFFFFFF)
+# define ARGON2_DEFAULT_FLAGS UINT32_C(0)
+# define ARGON2_FLAG_CLEAR_PASSWORD (UINT32_C(1) << 0)
+# define ARGON2_FLAG_CLEAR_SECRET (UINT32_C(1) << 1)
+# define ARGON2_BLOCK_SIZE 1024
+# define ARGON2_QWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 8
+# define ARGON2_OWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 16
+# define ARGON2_HWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 32
+# define ARGON2_512BIT_WORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 64
+# define ARGON2_ADDRESSES_IN_BLOCK 128
+# define ARGON2_PREHASH_DIGEST_LENGTH 64
+# define ARGON2_PREHASH_SEED_LENGTH 72
+
+# undef G
+# define G(a, b, c, d)                                                         \
+    do {                                                                       \
+        a = a + b + 2 * mul_lower(a, b);                                       \
+        d = rotr64(d ^ a, 32);                                                 \
+        c = c + d + 2 * mul_lower(c, d);                                       \
+        b = rotr64(b ^ c, 24);                                                 \
+        a = a + b + 2 * mul_lower(a, b);                                       \
+        d = rotr64(d ^ a, 16);                                                 \
+        c = c + d + 2 * mul_lower(c, d);                                       \
+        b = rotr64(b ^ c, 63);                                                 \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND
+# define BLAKE2_ROUND(v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11,        \
```

Comment:
Not sure about this name.. Its like a blake_round..

Reply:
Went with PERMUTATION_P (and PERMUTATION_P_{COLUMN,ROW}) in accordance with the Argon2 paper naming.